tests/common/freeze_time.py

```diff
@@ -2,82 +2,113 @@
+
+        # Ensure time has not been frozen from another coroutine
+        assert previous_task in (None, current_task)
+
+        # Set new context
+        __freeze_time_dict[dev_id] = (current_task, time)
+
+        restore_context[dev_id] = (previous_task, previous_time)
+
+    return restore_context
+
+
+def _unfreeze_devices(devices: list[LocalDevice], previous_context: FreezeContext) -> None:
+    for device in devices:
+        dev_id = device.device_id
+        previous_task, previous_time = previous_context[dev_id]
+        __freeze_time_dict[dev_id] = (previous_task, previous_time)
+
+        if previous_time is not None:
+            device.time_provider.mock_time(freeze=previous_time)
+        else:
+            device.time_provider.mock_time(realtime=True)
 
 
 @contextmanager
-def freeze_device_time(device: LocalDevice | DeviceID, current_time: DateTime | str):
+def freeze_devices_time(devices: list[LocalDevice], current_time: DateTime) -> Iterator[DateTime]:
     # Parse time
     if isinstance(current_time, str):
         [y, m, d] = current_time.split("-")
         current_time = DateTime(int(y), int(m), int(d))
 
-    # Get device id
-    if isinstance(device, LocalDevice):
-        device_id = device.device_id
-    elif isinstance(device, DeviceID):
-        device_id = device
-    else:
-        assert False, device
-
-    # Apply mockup (idempotent)
-    type(device).timestamp = _timestamp_mockup
-
-    # Save previous context
-    previous_task, previous_time = __freeze_time_dict.get(device_id, (None, None))
-
     # Get current trio task
     try:
         current_task = trio.lowlevel.current_task()
     except RuntimeError:
         current_task = None
 
-    # Ensure time has not been frozen from another coroutine
-    assert previous_task in (None, current_task)
+    restore_context = _freeze_devices(devices, current_task, current_time)
 
     try:
-        # Set new context
-        __freeze_time_dict[device_id] = (current_task, current_time)
         yield current_time
     finally:
         # Restore previous context
-        __freeze_time_dict[device_id] = (previous_task, previous_time)
+        _unfreeze_devices(devices, restore_context)
 
 
 __freeze_time_task = None
 
 
 @contextmanager
-def freeze_time(time: DateTime | str = None, device: LocalDevice | DeviceID | None = None):
-    mocks_stack = []
+def freeze_time(
+    time: DateTime | str | None = None,
+    devices: list[LocalDevice] | None = None,
+    global_freeze: bool = False,
+) -> Iterator[DateTime]:
+    mocks_stack: list[DateTime] = []
+    devices = devices or list()
 
     # Get current time if not provided
     if time is None:
         time = DateTime.now()
+    elif isinstance(time, str):
+        [y, m, d] = time.split("-")
+        time = DateTime(int(y), int(m), int(d))
 
     # Freeze a single device
-    if device is not None:
-        with freeze_device_time(device, time) as time:
+    if len(devices) > 0 and not global_freeze:
```

Comment:
And instead of the early return, it would be better more readable to do have a `if ... else ...`